IDEA/src/main/java/ksp/kos/ideaplugin/reference/context/LocalContext.kt

```diff
@@ -49,10 +60,20 @@ open class LocalContext @JvmOverloads constructor(
     protected open fun registerUnknown(type: ReferableType, name: String, element: Duality) {}
 
     protected fun addDefinition(type: ReferableType, name: String, element: Duality) {
-        getDeclarations(type)[name] = element
+        declarations.getOrPut(type, ::ScopeMap)[name] = element
+        // Make sure we can refer back to functions when they're used later in situations that work with variables.
+        if (type == ReferableType.FUNCTION) {
+            declarations.getOrPut(ReferableType.VARIABLE, ::ScopeMap)[name] = element
+        }
+        // Propagate globals back up so parent scopes can see them. Since the only way that imports can affect the
+        // current file's scope is via globals, we also pass those up.
+        // TODO - Maybe just put these on the file context?
```

Comment:
Since most (?) definitions are global, maybe we can create a [stub index](https://jetbrains.org/intellij/sdk/docs/basics/indexing_and_psi_stubs.html?search=index) with all (custom) global definitions?

Reply:
Ooh that's cool, I didn't know about that! How would you feel about me doing that in a followup, since this PR feels big enough and that looks like a decent chunk of work (and this isn't actually enabled yet so likely wouldn't have negative repercussions).

Reply:
My suggestion was definitely meant for a future PR :)